pkg/webhook/inject/inject.go

```diff
@@ -25,34 +25,55 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
-	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 
 	"github.com/chaos-mesh/chaos-mesh/pkg/annotation"
 	controllerCfg "github.com/chaos-mesh/chaos-mesh/pkg/config"
 	"github.com/chaos-mesh/chaos-mesh/pkg/metrics"
 	genericnamespace "github.com/chaos-mesh/chaos-mesh/pkg/selector/generic/namespace"
 	podselector "github.com/chaos-mesh/chaos-mesh/pkg/selector/pod"
 	"github.com/chaos-mesh/chaos-mesh/pkg/webhook/config"
+	"github.com/go-logr/logr"
 )
 
-var log = ctrl.Log.WithName("inject-webhook")
-
 var ignoredNamespaces = []string{
 	metav1.NamespaceSystem,
 	metav1.NamespacePublic,
 }
 
+type Injector struct {
+	logger        logr.Logger
+	decoder       *admission.Decoder
+	kubeclient    client.Client
+	cfg           *config.Config
+	ControllerCfg *controllerCfg.ChaosControllerConfig
+	Metrics       *metrics.ChaosControllerManagerMetricsCollector
+}
+
 const (
 	// StatusInjected is the annotation value for /status that indicates an injection was already performed on this pod
 	StatusInjected = "injected"
 )
 
+func NewInjector(logger logr.Logger, kubeclient client.Client, config *config.Config, controllerCfg *controllerCfg.ChaosControllerConfig, metrics *metrics.ChaosControllerManagerMetricsCollector) *Injector {
+	return &Injector{
+		logger:        logger,
+		kubeclient:    kubeclient,
+		cfg:           config,
+		ControllerCfg: controllerCfg,
+		Metrics:       metrics,
+	}
+}
+
 // Inject do pod template config inject
-func Inject(res *v1.AdmissionRequest, cli client.Client, cfg *config.Config, controllerCfg *controllerCfg.ChaosControllerConfig, metrics *metrics.ChaosControllerManagerMetricsCollector) *v1.AdmissionResponse {
+func (i *Injector) Inject(logger logr.Logger, res *v1.AdmissionRequest, cfg *config.Config, cli client.Client, controllerCfg *controllerCfg.ChaosControllerConfig, metrics *metrics.ChaosControllerManagerMetricsCollector) *v1.AdmissionResponse {
```

Comment:
I think the common serial of program is :
1. i := NewInjector(...)
2. i.Inject(...)

And then you can remove the first logger param here.